### PR TITLE
Implementing new trait for osu! API to return response body on errors

### PR DESCRIPTION
**Pull request by 112batuhan**
Implemented a new trait for request response struct.

Also implemented a new error struct. This error struct contains a struct variant with the status code and the body in String form if the error is caused by API request. This status code can be matched with the constants defined in the [documentation](https://docs.rs/reqwest/latest/reqwest/struct.StatusCode.html).
```
assert_eq!(
    err.status(),
    Some(reqwest::StatusCode::BAD_REQUEST)
);
```

### CODE REVIEW

**Review by Fursum — DISMISSED** [review]

**Comment by 112batuhan** [comment]
The error handling is not ready yet. I will work on the error conversion in `AppError`

**Comment by 112batuhan** [comment]
Completed @boraarslan

File: mi-osu-api/src/lib.rs [hunk 1]
```diff
@@ -14,3 +19,41 @@ pub use crate::beatmap::*;
 pub use crate::user::*;
 
 pub type ReqwestError = reqwest::Error;
+
+#[derive(Error, Debug)]
+pub enum OsuApiError {
+    #[error("Request failed with HTTP Status code.")]
+    HTTPError { body: String, error: StatusCode },
+    #[error("Request failed because of internal errors before or after the request.")]
```
Comment (boraarslan) [inline mi-osu-api/src/lib.rs:27]:
```suggestion
    #[error("An internal error has occurred.")]
```

File: mi-osu-api/src/auth.rs [hunk 2]
```diff
@@ -87,13 +87,15 @@ pub struct AuthResponseBody {
...
+) -> Result<AuthResponseBody, OsuApiError> {
     let response_result = client
         .post("https://osu.ppy.sh/oauth/token")
         .form(&body)
         .send()
         .await?;
-    response_result.json::<AuthResponseBody>().await
+    let response_result: Result<AuthResponseBody, OsuApiError> =
+        response_result.try_deserialising().await;
+    response_result
```
Comment (boraarslan) [inline mi-osu-api/src/auth.rs:98]:
```suggestion
    response_result.try_deserialising().await
```
Can you change this for all alike calls?

**Review by boraarslan — CHANGES_REQUESTED** [review]

**Review by boraarslan — APPROVED** [review]
Great job

**Pull request merged**